FIR checker/IDE: Report VAL_OR_VAR_ON_*_PARAMETER and enable RemoveValVarFromParameterFix

punzki opened the pull request:
Verified that this makes all `VAL_OR_VAR_ON_*_PARAMETER` match between FE 1.0 and FIR.

mglukhikh commented on compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:
```diff
@@ -65,6 +65,10 @@ object DIAGNOSTICS_LIST : DiagnosticList() {
+        val VAL_OR_VAR_ON_LOOP_PARAMETER by warning<KtParameter>(PositioningStrategy.VAL_OR_VAR_NODE)
+        val VAL_OR_VAR_ON_FUN_PARAMETER by warning<KtParameter>(PositioningStrategy.VAL_OR_VAR_NODE)
+        val VAL_OR_VAR_ON_CATCH_PARAMETER by warning<KtParameter>(PositioningStrategy.VAL_OR_VAR_NODE)
+        val VAL_OR_VAR_ON_SECONDARY_CONSTRUCTOR_PARAMETER by error<KtParameter>(PositioningStrategy.VAL_OR_VAR_NODE)
```
Just curious why have you decided downgrade LOOP / FUN / CATCH to warning and leave CONSTRUCTOR as error.

jsjeon commented on compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/expression/FirCatchParameterChecker.kt:
```diff
@@ -19,25 +20,30 @@ object FirCatchParameterChecker : FirTryExpressionChecker() {
+                reporter.reportOn(source, FirErrors.CATCH_PARAMETER_WITH_DEFAULT_VALUE, context)
+            }
+
+            if (source != null && source.hasValOrVar()) {
```
Nit: it's not about this PR, but rather usage/check of nullability of `source` in general (and hope this caught everyone's attention).

So... `reporter.reportOn` won't do anything if the given source is `null`. :) I remember I did add nullability checks almost everywhere in earlier stage of checkers, but recently not added anymore. Also, even in this checker, nullability of `catchParameter.source` wasn't checked at all. I'm not sure what is the convention.

punzki replied:
Good point, it isn't necessary and we should just continue of source is null.

punzki commented:
I also missed the error message (which contains the `val/var` keyword) for the diagnostics. The latest PR update has this.

tgeng commented:
Maybe we can add a test case in org.jetbrains.kotlin.checkers.FirKotlinHighlightingPassTestGenerated to cover diagnostic messages.

punzki commented:
> Maybe we can add a test case in org.jetbrains.kotlin.checkers.FirKotlinHighlightingPassTestGenerated to cover diagnostic messages.

Thanks for the suggestion! Added test in last commit. Glad you suggested it too, I was able to find and fix a bug in FE 1.0 plugin: https://youtrack.jetbrains.com/issue/KTIJ-13030

punzki commented:
@mglukhikh I added a test and in doing so, I found and fixed a bug in FE 1.0 plugin (see last commit). Thought you might be interested since you were the last person to change the logic there: https://github.com/JetBrains/kotlin/commit/0f9d31c9d1b536c301ce67c9cd4cb130e90f1bd1

tgeng commented on compiler/fir/checkers/checkers-component-generator/src/org/jetbrains/kotlin/fir/checkers/generator/diagnostics/FirDiagnosticsList.kt:
```diff
@@ -68,6 +69,18 @@ object DIAGNOSTICS_LIST : DiagnosticList() {
         val FLOAT_LITERAL_OUT_OF_RANGE by error<PsiElement>()
         val WRONG_LONG_SUFFIX by error<KtElement>(PositioningStrategy.LONG_LITERAL_SUFFIX)
         val DIVISION_BY_ZERO by warning<KtExpression>()
+        val VAL_OR_VAR_ON_LOOP_PARAMETER by warning<KtParameter>(PositioningStrategy.VAL_OR_VAR_NODE) {
```
Should this be an error instead of warning? Same below.

punzki replied:
According to the diagnostics list doc, we can downgrade to a warning. I believe it is @mglukhikh who put that note.

mglukhikh commented:
> I added a test and in doing so, I found and fixed a bug in FE 1.0 plugin (see last commit). Thought you might be interested since you were the last person to change the logic there: [0f9d31c](https://github.com/JetBrains/kotlin/commit/0f9d31c9d1b536c301ce67c9cd4cb130e90f1bd1)

Thank you, looks Ok.

darthorimar commented:
LGTM, will merge manually

darthorimar commented:
Thank you, merged manually
